test-adapter/src/main/scala/org/scalajs/sbttestadapter/TestAdapter.scala

```diff
@@ -0,0 +1,165 @@
+
+  private[this] val closing = new AtomicBoolean(false)
+
+  private[this] val nextRunId = new AtomicInteger(0)
+  private[this] val runs = TrieMap.empty[RunMux.RunId, Unit]
+
+  /** Creates an `sbt.testing.Framework` for each framework that can be found.
+   *
+   *  The returned Frameworks bind to this TestAdapter and are only valid until
+   *  [[close]] is called.
+   */
+  def loadFrameworks[TestFramework](
+      frameworkNames: List[List[String]]): List[Option[Framework]] = {
+    val mRunner = getRunnerForThread()
+
+    val infos = Await.result(
+        mRunner.com.call(JSEndpoints.detectFrameworks)(frameworkNames),
+        config.timeout)
+
+    infos.map(_.map(info => new FrameworkAdapter(info, this)))
+  }
+
+  /** Releases all resources. All associated runs must be done. */
+  def close(): Unit = {
+    if (!closing.getAndSet(true)) {
+      // Snapshot runs.
+      val seenRuns = runs.keySet
+
+      runners.values.foreach(_.com.close())
+      runners.values.foreach(_.runner.stop())
+      runners.clear()
+
+      runs.clear()
+      if (seenRuns.nonEmpty) {
+        throw new IllegalStateException(
+            s"close() called with incomplete runs: $seenRuns")
+      }
+    }
+  }
+
+  private[testadapter] def timeout(): Duration = config.timeout
+
+  private[testadapter] def runStarting(): RunMux.RunId = {
+    require(!closing.get(), "We are closing. Cannot create new run.")
+    val runId = nextRunId.getAndIncrement()
+    runs.put(runId, ())
+    runId
+  }
+
+  /** Called by [[RunnerAdapter]] when the run is completed. */
+  private[testadapter] def runDone(runId: RunMux.RunId): Unit = {
+    val old = runs.remove(runId)
+    require(old.nonEmpty, s"Tried to remove nonexistent run $runId")
+  }
+
+  private[testadapter] def getRunnerForThread(): ManagedRunner = {
+    // Prevent runners from being started after closing started.
+    // Otherwise we might leak runners.
+    require(!closing.get(), "We are closing. Cannot create new runner.")
+
+    val threadId = Thread.currentThread().getId()
+
+    // Note that this is thread safe, since each thread can only operate on
+    // the value associated to its thread id.
+    runners.getOrElseUpdate(threadId, startManagedRunner(threadId))
+  }
+
+  private def startManagedRunner(threadId: Long): ManagedRunner = {
+    // !!! DUPLICATE code with ScalaJSPlugin.makeExportsNamespaceExpr
+    val prefix = moduleKind match {
+      case ModuleKind.NoModule =>
+        ""
+
+      case ModuleKind.CommonJSModule =>
+        val moduleIdent = moduleIdentifier.getOrElse {
+          throw new IllegalArgumentException(
+              "The module identifier must be specified for CommonJS modules")
+        }
+        s"""require("${escapeJS(moduleIdent)}").""" // note the final '.'
+    }
+
+    val code = s"""${prefix}org.scalajs.testinterface.internal.startBridge();"""
+    val launcher = new MemVirtualJSFile(s"testMaster.js").withContent(code)
```

Comment:
`testMaster.js` does not feel appropriate anymore. Maybe `startTestBridge.js`?

(also the `s` interpolator is useless)